### PR TITLE
Scan boot classpath on JVM 9 and later

### PR DESCRIPTION
**Pull request by xxDark**
## What's new

Returns scanning of boot classpath on JVM 9 and later versions, checking of version is done by checking `java.class.version` system property

### CODE REVIEW

**Comment by Col-E** [comment]
Seems to not be working on Java 12. Does this work on Java 9?

```
Caused by: java.lang.IllegalArgumentException: Can not set final jdk.internal.loader.URLClassPath field jdk.internal.loader.BuiltinClassLoader.ucp to java.net.URLClassLoader
	... INTERNAL CALLS ...
	at java.base/java.lang.reflect.Field.set(Field.java:780)
	at me.coley.recaf.util.Classpath$ClassPathScanner.scanBootstrapClasspath(Classpath.java:235)
	at me.coley.recaf.util.Classpath$ClassPathScanner.scan(Classpath.java:206)
```

I noticed that this sets the field `URLClassPath ucp` to a `URLClassLoader`, but this fails because the types are incompatible.

**Comment by xxDark** [comment]
Should be fixed now, can't test right now.

**Comment by xxDark** [comment]
Also fixes #134

**Comment by Col-E** [comment]
Debugging on Java 12 currently, this does correctly scan the boot classpath on JDK 9+ (:+1: ) but I can't seem to verify that #134 is fixed. 

When running `mvn test` the `AutoCompleteTest` class is still skipped because `Classpath.isBootstrapClassesFound()` still returns false. 

image

The evaluation output should have names like `java.lang.Object`

**Comment by Col-E** [comment]
I can merge now if we want to tackle #134 later in a separate PR

**Comment by andylizi** [comment]
> Debugging on Java 12 currently, this does correctly scan the boot classpath on JDK 9+

Huh? When I tried it on Java 11, it didn't work because `bootstrapClasspath` is `null` (like I said above).

Did Java 12 change something? I only have 11 and 1.8 installed because they're the LTS versions.

**Comment by Col-E** [comment]
> Did Java 12 change something?

I was mistaken, see the edit.

**Comment by xxDark** [comment]
I guess because of Jigsaw module system boot classpath does not contain `java.*` and all other packages anymore, will look on what I can do with that

**Comment by xxDark** [comment]
Got it

**Comment by Col-E** [comment]
image
:ok_hand:

File: src/main/java/me/coley/recaf/util/Classpath.java [hunk 1]
```diff
@@ -138,38 +146,112 @@ public void scan(ClassLoader classLoader) {
...
+								.findVirtual(moduleClass, "getPackages", MethodType.methodType(Set.class));
+						MethodHandle getModule = lookup
+								.findVirtual(Class.class, "getModule", MethodType.methodType(moduleClass));
+						MethodHandle getLayer = lookup
+								.findVirtual(moduleClass, "getLayer", MethodType.methodType(layer));
+						MethodHandle layerModules = lookup
+								.findVirtual(layer, "modules", MethodType.methodType(Set.class));
+						MethodHandle unnamedModule = lookup
+								.findVirtual(ClassLoader.class, "getUnnamedModule", MethodType.methodType(moduleClass));
+						Set modules = new HashSet();
```
Comment (andylizi) [inline src/main/java/me/coley/recaf/util/Classpath.java]:
```suggestion
						Set<Object> modules = new HashSet<>();
```

Comment (andylizi) [inline src/main/java/me/coley/recaf/util/Classpath.java]:
```suggestion
						Set<Object> modules = new HashSet<>();
```

File: src/main/java/me/coley/recaf/util/Classpath.java [hunk 2]
```diff
@@ -138,38 +146,112 @@ public void scan(ClassLoader classLoader) {
...
+
+						Object ourModule = getModule.invoke(Classpath.class);
+						Object ourLayer = getLayer.invoke(ourModule);
+						if (ourLayer != null) {
+							modules.addAll((Set) layerModules.invoke(ourLayer));
+						}
+						modules.addAll(
+								(Set) layerModules
```
Comment (andylizi) [inline src/main/java/me/coley/recaf/util/Classpath.java]:
```suggestion
								(Collection<?>) layerModules
```

File: src/main/java/me/coley/recaf/util/Classpath.java [hunk 3]
```diff
@@ -138,38 +146,112 @@ public void scan(ClassLoader classLoader) {
...
+						for (Object impl : modules) {
+							for (String name : (Set<String>) getPackages.invoke(impl)) {
+								export.invoke(impl, name);
+							}
+						}
+
+						Class<?> reflectionClass = Class.forName("jdk.internal.reflect.Reflection");
+						MethodHandle getMapHandle = lookup.findStaticGetter(reflectionClass, "fieldFilterMap", Map.class);
+						// Map is immutable, thanks Oracle
+						Map<Class<?>, Set<String>> fieldFilterMap = new HashMap<>((Map<Class<?>, Set<String>>) getMapHandle.invokeExact());
```
Comment (andylizi) [inline src/main/java/me/coley/recaf/util/Classpath.java]:
```suggestion
						Map<Class<?>, ?> fieldFilterMap = new HashMap<>((Map<Class<?>, ?>) getMapHandle.invokeExact());
```

They changed some code in `Reflection` in JDK 12: [before](https://github.com/openjdk/jdk11u/blob/71f32ecc0fa1d0c552c40b1222412a70678dea19/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L46-L58) and  [after](https://github.com/openjdk/jdk12u/blob/fca351466233b0ac5867f3cd690897df3d707726/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L45-L63), so 

1. The generic signature of `fieldFilterMap` should be `Map<Class<?>, ?>` for compatibility between `Map<Class<?>, String[]>` and `Map<Class<?>, Set<String>>` (we are only using the key anyway).
2. What's more, it's altogether unnecessary to do this before Java 12 - older versions didn't filter out `Field` access.

File: src/main/java/me/coley/recaf/util/Classpath.java [hunk 4]
```diff
@@ -138,38 +146,112 @@ public void scan(ClassLoader classLoader) {
...
+						listReader.setAccessible(true);
+						Field bootModulesField = bootLoader.getClass().getSuperclass().getDeclaredField("packageToModule");
+						bootModulesField.setAccessible(true);
+						Collection<?> packageToModule = ((Map<String, ?>) bootModulesField.get(null)).values();
+						for (Object module : packageToModule) {
+							// Scan it!
+							Object ref = mref.invoke(module);
+							Object reader = openReader.invoke(ref);
+							Stream<String> list = (Stream<String>) listReader.invoke(reader);
+							list.parallel().filter(s -> s.endsWith(".class"))
```
Comment (andylizi) [inline src/main/java/me/coley/recaf/util/Classpath.java]:
```suggestion
							list.filter(s -> s.endsWith(".class"))
```

[Should I always use a parallel stream when possible? - Stack Overflow](https://stackoverflow.com/a/20375622/6238618)

**Comment by Col-E** [comment]
Any more commits to add or are we all ready to merge?

**Comment by xxDark** [comment]
> 
> 
> Any more commits to add or are we all ready to merge?

Ready

**Comment by Col-E** [comment]
Thank  for the PR :+1:

**Pull request merged**